modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,472 @@
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::{prelude::Vec, vec};
+
+/// The finality proof used by the pallet.
+///
+/// For a Substrate based chain using Grandpa this will
+/// be an encoded Grandpa Justification.
+pub type FinalityProof<'a> = &'a [u8];
+
+/// Errors which can happen while importing a header.
+#[derive(Debug, PartialEq)]
+pub enum ImportError {
+	/// This header is older than our latest finalized block, thus not useful.
+	OldHeader,
+	/// This header has already been imported by the pallet.
+	HeaderAlreadyExists,
+	/// We're missing a parent for this header.
+	MissingParent,
+	/// We were unable to prove finality for this header.
+	UnfinalizedHeader,
+	/// We failed to verify this header's ancestry.
+	AncestryCheckFailed,
+	/// The header is missing digests related to consensus events.
+	///
+	/// This will typically have to do with missing authority set change signals.
+	MissingConsensusDigest,
+}
+
+/// A trait for verifying whether a header is valid for a particular blockchain.
+pub trait ChainVerifier<S, H> {
+	/// Import a header to the pallet.
+	fn import_header(storage: &mut S, header: &H, finality_proof: Option<FinalityProof>) -> Result<(), ImportError>;
+
+	/// Verify that the given header has been finalized and is part of the canonical chain.
+	///
+	/// Returns a list of headers which got finalized by the given header.
+	fn verify_finality(storage: &mut S, header: &H, proof: FinalityProof) -> Result<Vec<H>, ImportError>;
+}
+
+/// Used to verify imported headers and their finality status.
+#[derive(Debug)]
+pub struct Verifier;
+
+impl<S, H> ChainVerifier<S, H> for Verifier
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	fn import_header(storage: &mut S, header: &H, finality_proof: Option<FinalityProof>) -> Result<(), ImportError> {
+		let highest_finalized = storage.best_finalized_header().expect("TODO");
+		if header.number() < highest_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = storage.get_header_by_hash(*header.parent_hash());
+		if parent_header.is_none() {
+			return Err(ImportError::MissingParent);
+		}
+
+		// A block at this height should come with a justification and signal a new
+		// authority set. We'll want to make sure it is valid
+		//
+		// This defaults to 0, should it maybe be an Option?
+		let scheduled_change_height = storage.scheduled_set_change().height;
+		if *header.number() == scheduled_change_height && finality_proof.is_some() {
+			// Maybe pass the scheduled_change in here so we don't have to query storage later
+			let finalized_headers = Self::verify_finality(
+				storage,
+				header,
+				finality_proof.expect("Checked for `finality_proof` before entering if-block"),
+			)?;
+
+			// TODO: Would need to prune blocks from the non-canonical chain at some point
+			for header in finalized_headers.iter() {
```

Comment:
`verify_finality` only returns ancestors? Probably should also write current header too?

Reply:
I didn't want to have any header writes in `verify_finality`. I wanted `import_header` to be the only thing doing as a way to minimize side-effects (or at least have one clear location where writes are taking place)

Reply:
Are you answering my question here? :) What I said is that `finalized_headers` is missing `header` (i.e. the header which you're importing now).

Reply:
> Are you answering my question here? 

I thought you were suggesting that `Verifier::verify_finality` should be writing to storage in general.

> What I said is that finalized_headers is missing header (i.e. the header which you're importing now)

I don't think so. If you take a look at the test `correctly_verifies_and_finalizes_chain_of_headers()`, which calls  `Verifier::import_header` with a `header`, you can see that `header` gets written to storage correctly. Maybe the name `finalized_headers` is a little misleading here?

Reply:
Ok, I think you're right - sorry for confusion :)